foundationdb/src/timekeeper.rs

```diff
@@ -0,0 +1,96 @@
+//! There is a key range called TimeKeeper in the system key space which stores a rolling history window
+//! of time to version mappings, with one data point every 10 seconds.
+//! It is not exposed via any user-facing API, though of course the data can be read by a user.
+//! It is not an official database feature and should not be relied on for anything where accuracy
+//! is critical as nothing prevents or detects system clock skew on the FDB process logging these data points.
+//!
+//! TimeKeeper is used by backup and restore to convert timestamps to approximate versions and versions
+//! to approximate timestamps to make reasoning about backup data and restore operations easier.
+//! Lookups work by finding the nearest value for the query version or timestamp, taking the equivalent other value,
+//! and then adding an adjustment estimate based on 1 million versions per 1 second.
+//! This logic accounts for arbitrary version advancement due to recovery, DR switch operations, or any other reason.
+//!
+//! [source](https://forums.foundationdb.org/t/versionstamp-as-absolute-time/2442/3)
+
+use crate::future::FdbValue;
+use crate::options::TransactionOption;
+use crate::{FdbBindingError, FdbResult, KeySelector, RangeOption, Transaction};
+use foundationdb_tuple::{pack, unpack};
+use futures::StreamExt;
+
+/// Timekeeper keys are stored in a special keyspace
+/// Can be found in the [Java implementation](https://github.com/FoundationDB/fdb-record-layer/blob/main/fdb-extensions/src/main/java/com/apple/foundationdb/system/SystemKeyspace.java#L80)
+const TIME_KEEPER_PREFIX: &[u8] = b"\xff\x02/timeKeeper/map/";
+
+/// Flavor about the mode of scanning
+pub enum HintMode {
+    /// The read version is ensure to be before the timestamp
+    BeforeTimestamp,
+    /// The read version is ensured to be after the timestamp
+    AfterTimestamp,
+}
+
+/// Try to get a version ID closer as possible as the asked timestamp
+///
+/// If no result are found, either your timestamp is in the future of the
+/// Timekeeper or the data as been rolled by fresh ones.
+///
+/// The layout os follow:
+///
+/// TIME_KEEPER_PREFIX/timestamp1 => read_version1
+/// TIME_KEEPER_PREFIX/timestamp2 => read_version2
+/// TIME_KEEPER_PREFIX/timestamp3 => read_version3
+///
+/// Each key are associated to a pack read version on 8 bytes
+/// compatible with an i64.
+pub async fn hint_version_from_timestamp(
+    trx: &Transaction,
```

Comment:
non-blocking: taking a transaction here means that we won't be able to call this API from db.run. Unfortunately we don't have a trait or something to mask this complexity. We will deal with this later.

Reply:
I was plannining to add a public method to the RetryableTransaction which calls my logic
```rust
impl RetryableTransaction {
    pub async fn hint_version_from_timestamp(
       &self, 
       timestamp: u64, mode: HintMode,
       snapshot: bool,
     ) -> Result<Option<u64>, FdbBindingError> {
           hint_version_from_timestamp(&self, timestamp, mode, snapshot).await
     }
}
````

Reply:
The API is fully compatible with db.run

```rust
database
    .run(|trx, _| async move {
        hint_version_from_timestamp(&trx, now, HintMode::AfterTimestamp, true)
            .await
            .unwrap();
        Ok(())
    })
    .await
    .unwrap();
```